### ISSUE TITLE
Aria-AT Wireframes: Test Management, Initiating a Test Cycle, Test Cycle Summary Page, Executing Tests, Raw Test Results and Draft Results

### ISSUE THREAD

**Issue by isaacdurazo** [w3c/aria-at#132]
Wireframes Wiki Page: https://github.com/w3c/aria-at/wiki/Aria-AT-Wireframes
Use Cases in issue #102

# Test Management
This page is accessible from the main menu. There are two sections in this page: 1) Initiate a Test Cycle and 2) Test Cycle Status

## 1. Initiate a Test Cycle 
Here, with the click of a button, a new Test Cycle initiation can be executed. This will take the user to another page where the Test Cycle can be configured. 

## 2. Test Cycle Status 
This is a table that displays the Test Cycles that have been initiated. For every Test Cycle there's a row in the table. This table has the following columns: 
   1. Test Cycle: It contains a link with the name given to the Test Cycle when it was configured. Clicking this will take the admin to the Test Cycle Summary Page.
   2. Date: The date in which the Test Cycle was initiated.
   3. Status: In progress, raw results available, draft results available and reviewed results available.
   4. More Details: The cells of this table contain buttons/links to raw results, draft results and reviewed results.

![Test Management](https://user-images.githubusercontent.com/1379244/77789855-11051980-7021-11ea-93e6-d3b85803b4e8.png)

# Initiating a Test Cycle
This page is generated from the Test Management Page. There are 3 main sections in this page: 1) Test Cycle Configuration, 2) Test plans that are missing assignees and 3) Test Plans Available

## 1. Test Cycle Configuration 
This is a table where the admin can choose a custom name for the cycle, initiate a cycle from a specific commit and choose specific Assistive Technology and Browser versions. This table has the following characteristics:

   1. Test Cycle name: This is a text field where a custom name can be added
   2. Specific commit: From a dropdown, the admin can choose a specific commit
   3. AT and Browser versions: This table has as many rows as AT and browser combinations are available. There's a row per AT and Browser combination. It has the following columns:
      1. Assistive Technology: Each cell has a dropdown that contains the available AT released versions such as: beta, stable, dev, etc
      2. Assistive Technology Version Number: This cell contains a text field where the admin can input a specific version number for the AT
      3. Browser: Each cell has a dropdown that contains the available browser stage released versions such as: beta, stable, etc
      4. Browser Version number: This cell contains a text field where the admin can input a specific version number for the Browser.

## 2. Test plans that are missing assignees
This table displays test plans that are missing assignees. When a Test Cycle configuration is initiated, testers are assigned automatically by the system depending on the assistive technologies and browsers they can use and have recorded in their profile page. It is important to mention that even if the admin leaves test plans that are missing testers, the Test Cycle can still be initiated. Tests without a specific assignee will be available to all testers in the system. This table has the following characteristics:
 1. A status message at the top with a summary of the total number of assignees that are missing.
 2. The following columns:
    1. Test Plans:  Where the admin can see test plans that are missing assignees
    2. Assistive Technology and Browser: Where the admin can see the AT/Browser combination under a Test plan that require assignees. 
    3. Testers: Where the admin can assign testers from a dropdown.

## 3. Test Plans Available
This section include all available Test Plans in the system. The admin can decide to submit them all as part of a Test Cycle or choose a subset of these. Each Test Plan has its own table with all the available AT and Browser combinations available in the system as well as the testers assigned to execute certain Test plans with certain AT and Browser combinations. This table has the following characteristics:
 1. Test plan name: This is the title for the test plan. There's a checkbox next to it to select the test plan as part of the Test Cycle. All test plans are selected by default.
 2. The Test Plan table has the following columns:
    1. AT and Browser: This column lists the AT and Browser combinations available in the system. One combination is displayed by row. Each includes a checkbox to make a custom selection of AT/Browsers per Test Plan.
    2. Tester 1: This includes a dropdown where tester 1 is automatically selected by the system. The admin has the ability to override this selection choosing a different tester from the dropdown.
    3. Tester 2: Same characteristics of Tester 1.

At the bottom of this page, there's a UI component that contains a checkbox to check/uncheck all test plans and a button to initiate the Test Cycle.

![Initiating a Test Cycle](https://user-images.githubusercontent.com/1379244/77789911-2417e980-7021-11ea-8b12-855b14c7dcc8.png)

# Test Cycle Summary Page
This page shows a high level summary of the status of a Test Cycle and is accessible from the Test Management page by clicking the name of the Test Cycle from the Test Cycle Status table. There are 3 sections in this page: 1) Summary of configurations, 2) Test plans that still need assignees and 3) Test Plans included in the Test Cycle

## 1. Summary of configurations
This section displays the summary of configurations chosen while initiating the Test Cycle: Test Cycle name, git commit, specific versions of AT and Browsers if any were selected.

## 2. Test plans that still need assignees
This table is like the one we have in the "Initiating a Test Cycle" page. The admin can still select testers if needed and the system will update this automatically.

## 3. Test Plans included in the Test Cycle
This table is like the one we have in the "Initiating a Test Cycle" page, where the admin configures and initiates a Test Cycle, with two addition: 
 1. A message with a summary of the progress for the Test Plan
 2. A progress column that shows the percentage of progress and how many tests have been executed and how many are left, e.g. 6/16.

At the bottom of this page, there's a UI component that contains a checkbox to check/uncheck all test plans and a button to review the raw results for the Test Cycle.

![Test Cycle Summary Page](https://user-images.githubusercontent.com/1379244/77789960-33973280-7021-11ea-8106-1ef114c513b0.png)

# Executing Tests
This page is also known as the "Testers Queue". It shows the Test Plans that were defined by the admin when created the Test Cycle. These Test Plans are grouped by AT and Browser combination and for every combination there is a table that includes the test plans, testers, progress and the option to initiate the execution of all test plans belonging to said AT/Browser combination. This table has the following columns:
  1. Test Plan: where all the Test plans available are listed.
  2. Tester 1: This include the name of the tester, unless no one has been assigned. In that case, a button to assign yourself as tester will be displayed
  3. Tester 2: This includes the name of the other tester that is executing the same Test Plan using the same AT and Browser, unless there's no assignee. In that case the text: "No assignee" will be displayed.
  4. Progress: It shows the percentage of progress and how many tests have been executed and how many are left, e.g. 6/16.

Once the tester has decided what AT, Browser and Test to execute, the process looks pretty similar to what we currently have in the prototype. A few changes have been made:

  1. Skip a test: when the tester decides to skip a test, a dialog will open. This dialog contains details about the test being skipped as well as a button to open a Github issue to describe elaborate on the reason for skipping the test. Once the Github issue has been created, the tester can now go back to the application and from the dialog, click continue to move to the next test in the queue.
  2. Save progress: if the tester needs to pause the execution of a test, they can simply click this button and leave the application.

Note: the process for executing Tests is the same we have in the prototype.

![Executing Tests](https://user-images.githubusercontent.com/1379244/77789973-398d1380-7021-11ea-8823-a9079ffd7bf6.png)

# Raw Test Results
This page shows a summary of the raw results of a Test Cycle that was executed by a tester. There are 2 main sections in this page: 1) Summary of configurations 2) All the Test Plans executed as part of this cycle. 

## 1. Summary of configurations
This section displays the summary of configurations chosen while initiating the Test Cycle: Test Cycle name, git commit, specific versions of AT and Browsers if any were selected.

## 2. Test Plans
Each Test plan section has two main sections: 1) Tests with conflicting results and 2) Successful Test results. 

### 1. Tests with conflicting results
A table with a summary of the tests that have conflicting test results and need to be evaluated and approved by the admin. This table has a message with a summary at the top that shows the total number of tests with conflicting results in the test plan and the following characteristics and columns:

  1. Test name: The name of the test, e.g. "Navigate into a checkbox group in interaction mode.". Next to it in the next column, sub-rows for every AT and Browser are displayed. 
  2. AT and Browser: Each AT and Browser row in this table has two sub-rows for every tester so results can be compared. In addition to the AT and Browser name, there's a "details" link to see the test failures table, which looks the same as the one we have in the prototype, with the exception of this being now displayed in a dialog. 
  3. Testers: The name of the tester that executed the test
  4. Required:  summary of results. Same format we have in the prototype
  5. Optional:  summary of results. Same format we have in the prototype
  6. Unexpected Behavior
  7. Action: the cells in this column have a radio button for every tester so the Admin can decide which test is correct.

Once the admin has decided what results are correct, they can click the "apply changes" button, which will subsequently update the results in the tables that have successful test results.

### 2. Successful Tests results
There are two ways to visualize test results in this page: 1) grouping the test plans by AT and Browser combination and 2) grouping by tests. Either option can be chosen by using a dropdown menu

#### Grouping by AT and Browser combination
When grouping by AT and Browser combination, a table for every combination will be displayed. Each of these tables has the following characteristics and columns:

  1. Test: The name of the test e.g. "Navigate into a checkbox group in interaction mode." and a "details" link to see the test failures table, which looks the same as the one we have in the prototype, with the exception of this being now displayed in a dialog.
  2. Testers: the name of the two testers that executed the test
  4. Required: summary of results. Same format we have in the prototype
  5. Optional: summary of results. Same format we have in the prototype
  6. Unexpected Behavior

#### Grouping by tests
When grouping by tests, one single table is displayed per Test Plan. This is a compressed way to review results. This table has the following characteristics and columns:

  1. Test: The name of the test e.g. "Navigate into a checkbox group in interaction mode." next to it, in the next column, sub-rows for every AT and Browser are displayed.
  2. AT and Browser: In addition to the AT and Browser name, there's a "details" link to see the test failures table, which looks the same as the one we have in the prototype, with the exception of this being now displayed in a dialog. 
  3. Testers: Name of both testers that executed the tests.
  4. Required: summary of results. Same format we have in the prototype
  5. Optional: summary of results. Same format we have in the prototype
  6. Unexpected Behavior

At the bottom of the page, there's a button to generate Draft Results

![Raw Results](https://user-images.githubusercontent.com/1379244/77790005-4873c600-7021-11ea-9ef9-02490c8c54ef.png)

# Draft Results
This page shows a summary of the results of a Test Cycle that was executed by a tester. It is generated from the Raw results and the only difference from the Raw Results page is that the "Tests with conflicting results" table is no longer displayed. At the bottom of this page, the admin is able to publish the results.

![Draft Results](https://user-images.githubusercontent.com/1379244/77790035-545f8800-7021-11ea-9c6a-35bf636f97dd.png)

**Comment by Yohta89** [comment]
One quick clarification for the terminology.
Under 'Raw Test Results' heading, there were several mentions of conflicting results, which was slightly unclear to me.

Is it conflicting results among multiple testers, or is there any cases generating conflicting results by a single tester? Since the latter seems unlikely, we might want to clarify what's conflicting results/what would be the cases of conflicting results.

**Comment by mcking65** [comment]
Isaac, I love, love, love the descriptions of the UI! Fabulous work!!!!! Very nicely done!!!!

I've read it all but not finished thinking about it all. Here are some initial thoughts.

Initiating a Test Cycle > 2. Test plans that are missing assignees:

>This table displays test plans that are missing assignees. When a Test Cycle configuration is initiated, testers are assigned automatically by the system depending on the assistive technologies and browsers they can use and have recorded in their profile page.

Giving this a second thought, we should discuss whether we want to invest in auto-assignment. Perhaps we could simplify by defaulting to starting cycles with all test plan runs in an up for grabs state and individual testers have a view of the queue that is filtered by their profile settings.

When an admin manually assigns a specific tester to a test plan run, the list of available testers could also be filtered by the tester's profile.

When it comes to auto-assignment, I am concerned about how load would be distributed across available testers, especially if each tester has a different level of availability capacity. It might get complicated.

At the same time, I can imagine having some planning meetings or discussions with testers where it is useful to assign certain runs to certain testers. So, it might be useful to have a checkbox column that enables a set of rows to be selected. Then, outside the table, have an assign button that would show a combobox populated with testers qualified to execute all the selected runs. After assignment to the chosen person, any selected rows that still have only one tester assigned would remain checked in the event that the admin wants to assign that same set of rows to another person. There should be buttons for select all and unselect all rows.

>The Test Plan table has the following columns: 
>i. AT and Browser: This column lists the AT and Browser combinations available in the system. One combination is displayed by row. Each includes a checkbox to make a custom selection of AT/Browsers per Test Plan. 
>ii. Tester 1: This includes a dropdown where tester 1 is automatically selected by the system. The admin has the ability to override this selection choosing a different tester from the dropdown. 
>iii. Tester 2: Same characteristics of Tester 1. 

Instead of tester 1 and tester 2, consider having a column for "Assigned testers". Put a list in each cell with 3 entries: tester 1, tester 2, and "Add another tester button".

Or, give ability to add another column -- that might be harder.

We should be careful to not limit the UI design to accommodate only 2 testers per test run plan. We've decided to do this for the initial process, but we might learn that we want to run things differently over time. Maybe for some test plans we want 3. This change would cascade into subsequent screens and tables.

**Comment by zcorpan** [comment]
> Action: the cells in this column have a radio button for every tester so the Admin can decide which test is correct.

From the [telecon discussion yesterday](https://www.w3.org/2020/04/01-aria-at-minutes.html), @mcking65 suggested that instead of the test admin deciding on their own which result is correct, the system would facilitate discussion in a GitHub issue for conflicting results, and the testers can go back and re-run / change their raw test results. When the 2 raw test results agree, the draft test results can be published.

**Comment by isaacdurazo** [comment]
Based on the feedback for how to handle assignments and how to deal with conflicting results, I've made changes to the following wireframes: 1) Initiating a Test Cycle, 2) Test Cycle Summary Page, 3)  Executing Tests and 4) Raw Results.

Note: This response includes only details about the changes. I will be updating the wiki page with a complete new version of all the wireframes early next week.

# Initiating a Test Cycle
Two sections in this page have been modified: 1) Test plans that need assignees and 2) Test Plans Available

##  Test plans that need assignees
This section has been removed. Testers will no longer be assigned automatically by the system,  instead, the admin will do this manually.

## Test Plans Available
The following changes have been made to this section:
  1. One of the two columns for testers has been removed, instead, we will display all assignees in one column.
  2. The dropdown menus for each row in the Testers column have been changed to display the assignee's name as plain text.
  3. A new component has been attached to the bottom of the test plan table were the admin can now handle assigning multiple testers at the same time based on the rows that the admin has selected. This component has a dropdown that when opened, displays the list of testers that are qualified to execute all the selected runs. From this dropdown the admin can assign and un-assign testers.

![Initiating a Test Cycle](https://user-images.githubusercontent.com/1379244/78413286-91042400-75cb-11ea-9ff0-f98d1d44df1c.png)

# Test Cycle Summary Page
Two sections in this page have been modified: 1) Test plans that need assignees and 2) Test Plans Available

##  Test plans that need assignees
The same changes made in the "Initiating a Test Cycle" page have been made here. This section has been removed

## Test Plans Available
The same changes made in the "Initiating a Test Cycle" page have been made here: one of the two columns for testers has been removed, the dropdown menus in the testers column have been changed to display the assignee's name and the new component to assign multiple testers at the same time has been attached at the bottom of every test plan's table.

![Test Cycle Summary Page](https://user-images.githubusercontent.com/1379244/78413349-d163a200-75cb-11ea-983c-919b0f7f0f76.png)

# Executing Tests
Only the Tester's queue's tables have been changed in this page. Here are the changes:
  1. One of the two columns for testers has been removed, instead, we will display all assignees in one column.
  2. Checkboxes have been added to each row add the ability of single test plan execution in case of needing to redo tests due conflicting results between testers.
  3. A Status column has been added where the following status could be displayed: Not started, in progress, complete and complete with conflicts 

![Executing Tests](https://user-images.githubusercontent.com/1379244/78414031-0c1b0980-75cf-11ea-90e8-d7b2800e240f.png)

# Raw Test Results
Only the "Tests with conflicting results" section has been changed in this page. Since the admin is no longer in charge of choosing which tests are correct, the "Actions" column has been removed.

![Raw Results](https://user-images.githubusercontent.com/1379244/78413638-353a9a80-75cd-11ea-8ce0-0d7bf676b4cb.png)

**Comment by mcking65** [comment]
@isaacdurazo, Here's the rest of my feedback on the design for these screens.

#### General feedback

Per our discussion Thursday, if you don't update the images, please update descriptions in any place where you change the design based on feedback. It is good for the team to understand, then that the text is the plan of record if it differs from the image.

On all the pages with sections, I assume there are headings that name the sections. If not, please include headings in the design.

Based on last Wednesday's discussion about how to resolve conflicts, I think we need further discussion of process before designing raw test results and draft results pages. In fact, we may not need those pages at all. I am thinking we may want to build their functionality into the runner itself.

High-level summary of what this might look like:
1. Tester 1 completes test plan run X
1. Tester 2 completes filling in the form for test 1 in test plan run X and presses the review button in the runner.
1. The system compares tester 2 results to tester 1 results, displays the review dialog. If there are any conflicts, this dialog could also include information about conflicts.
1. If there are conflicts, tester 2 can ignore them and save the conflicting results, return to the test 1 page and re-do some of the work, or decide not to save any results for test 1 and continue to test 2, or discontinue testing. The tester may decide to stop testing and ask questions outside the system.
1. The tester could optionally raise an issue about a specific test. A tester may tell the system that the issue is a blocking issue. If there were conflicts, the issue summary would include the conflicts. We need to discuss the possible resolution paths, e.g., tester 1 re-tests, tester 2 retests, either tester is replaced with a different tester who re-runs the tests, the test is changed. A future system requirement could be that a test cycle can be edited to change the version of the tests, and any test that has changed would get set back to the not-started state.
1. The conflicts are resolved when there are 2 or more equivalent test results for each test.

In other words, instead of dedicated pages for working with raw results, we might be able to do it all in the runner. This would require the ability to load existing results into the runner when appropriate.

And, instead of waiting till a test plan run is complete to compare results, we could do it in real time as each result is entered into the system and make it part of review by the tester.

This also requires thinking carefully about the use cases in the runner for all types of users. I've written some detail about that below.

#### Defining terms and process

To specify the process, system, and UI requirements for a process like this, we need to spell out the process in more detail. And, to do that, I think we need some more terminology.

Test Result: Data collected from 1 tester from all fields in 1 test in a test plan run, i.e., outputs, assertions, and other issues for each command executed for the test. Let's say we are talking about a test plan run for checkbox where the plan version is V1, the AT is JAWS 2020, and the browser is Chrom 80. Let's say there are 16 tests, in a specific sequence in this plan. So, test result 1 from tester 1 for this test plan run would contain all the assertion results, outputs, etc. for test 1 in V1 of the test plan using JAWS 2020 with Chrome 80. Since there are 16 tests, if 2 testers were required to execute the tests, when all testing is done for that run, there would be 16 test results from tester 1 and 16 results from tester 2.

When test results for a given test in a given test plan run are being compared, the comparison can yield either of two outcomes:
* Equivalent: Outputs (less trimmed leading and trailing whitespace) and assertions are identical. And: either both results are free of other issues, or other issues are present for the same commands in both results. If other issues are present for the same commands, they do not need to be identical.
* Conflicting: They are not equivalent.

Using this terminology, I would like to propose some state definitions for tests and test plan runs.

Error states for a test in a test plan run:
* Blocked: an issue has been raised that the tester says blocks completion of the test.
* Conflicted: Conflicting test results have been submitted.
* Conflicted and blocked: Conflicting results have been submitted and a blocking issue has been raised.

Test plan run states:
* Not started: There are no test results for the test run plan.
* Complete: All the following are true:
   * At least two testers have submitted results for all tests.
   * All assigned testers have submitted results for all tests. Note that a test plan run that met all other conditions for being complete could become incomplete if a test admin assigned a 3rd or 4th tester to it.
   * All submitted test from all testers are equivalent.
* In progress: There is at least 1 test result submitted. If two or more test results have been submitted for the same test, the results are equivalent.
* In progress with conflicts: At least one test has 2 test results that are not equivalent and at least one test is missing a result that is necessary to meet the conditions for a complete status.
* Pending conflicts: All conditions for being complete have been met except one or more tests have results that are not equivalent.
* Blocked: Has at least one blocking issue. It may or may not have some or all results with or without or conflicts.

Now, here's feedback on the following pages with all this in mind.

### Test cycle summary 

On the Test Cycle Summary Page, if there are any test plans excluded from the cycle, there should be a section for test plans excluded from the test cycle. There should be a way for the admin to include any of the excluded plans.

#### Executing Tests

>This page is also known as the "Testers Queue".

I like the name "Test Queue" -- it's short and clear. It works for all users whether they are admins, testers, or general public.

I don't know if or where we have defined a top-level or banner navigation that would be at the top of every page. I am picturing we have the following on every page:
1. W3C Home (logo)
1. ARIA-AT Home
1. Reports
1. Test Queue
1. Test Admin
1. Users
1. Log in/Log out button: log in if user is not authenticated, log out if authenticated.

We haven't talked about a home page. We may not need it in V1. Although, it might be very simple to include, and it could be quite valuable. Ideally, the main content of the page would be defined in a markdown file that is easily changed without rebuilding the app.

>It shows the Test Plans that were defined by the admin when created the Test Cycle. These Test Plans are grouped by AT and Browser combination

I assume each grouping is a heading. Perhaps it should be collapsable, e.g., a clickable heading made like `<h3><button aria-expanded="false">JAWS with Chrome - 3 not started, 1 in progress, 3 finished</button></h3>`

Are we assuming we can have only 1 cycle in progress at a time? That might be OK for now. However, I can imagine scenarios in the future where we are running a cycle on some beta software and a cycle for some production software at the same time. In that case, there would be a heading for each cycle.

We should probably start now with an H1 for the page title, e.g., `<h1>Test Queue</h1`. Then and h2 for the Test cycle, e.g., `<h2>June 2020 Desktop Screen Reader Report Cycle</h2>`. Then, an H3 for each AT/Browser combination included in that cycle.

It could be very helpful to have some filters at the top of the test queue for each test cycle, that is under the H2 have filters for:

1. All test plan runs
1. Incomplete test plan runs (in progress, pending conflicts, blocked)
1. Runs with conflicts
1. Blocked runs
1. If the user is a logged in tester, My Work Queue (assigned plans where the user has not submitted all results + available plans where the user has not submitted results)

> and for every combination there is a table that includes the test plans, testers, progress and the option to initiate the execution of all test plans belonging to said AT/Browser combination. This table has the following columns:
>1. Test Plan: where all the Test plans available are listed. 
>2. Tester 1: This include the name of the tester, unless no one has been assigned. In that case, a button to assign yourself as tester will be displayed
>3. Tester 2: This includes the name of the other tester that is executing the same Test Plan using the same AT and Browser, unless there's no assignee. In that case the text: "No assignee" will be displayed. 
>4. Progress: It shows the percentage of progress and how many tests have been executed and how many are left, e.g. 6/16. 

Are tester names in the table going to be links that open their profile? I think they should be.

Per prior feedback, combine columns 2 and 3 into a "testers" column. It may list multiple testers per plan.

I propose that column 1 should be a link that opens the test plan run:
* If the current user is assigned, any results previously saved by that user are loaded/displayed. The first test in the sequence with no results is displayed. If all tests have results, the first test is displayed.
* If the current user is a tester who is qualified to run the plan but is not assigned, prompt to ask "Would you like to be assigned to the plan "{planName}?" with buttons for "Yes, assign me" and "No, I am just browsing the plan".
* (Stretch goal) If the user is an Admin and results have been saved, display results from the tester that started the plan first. At the top of each page, show who saved the results and when. If more than one tester has saved results, at the top of the page, provide a selector for choosing which tester's results are displayed.
* If the current user is not an admin and not assigned, do not show any results. Some of the buttons, e.g., Review and Submit, will be hidden or disabled as described below.

I propose another column that appears for testers and admins containing an actions button that opens a menu. This menu could include:
1. View tests with conflicts: Opens the runner but showing only the tests with conflicting results
1. View tests with blocking issues: Opens the runner but showing only tests where blocking issues have been raised.

>Once the tester has decided what AT, Browser and Test to execute, the process looks pretty similar to what we currently have in the prototype. A few changes have been made:
>1. Skip a test: when the tester decides to skip a test, a dialog will open. This dialog contains details about the test being skipped as well as a button to open a Github issue to describe elaborate on the reason for skipping the test. Once the Github issue has been created, the tester can now go back to the application and from the dialog, click continue to move to the next test in the queue. 
>2. Save progress: if the tester needs to pause the execution of a test, they can simply click this button and leave the application. 

In the prototype, we  have the following buttons:
1. Review Results
2. Redo Test
3. Skip Test
4. Submit Results

I think we need to support the following use cases:
1. Tester who is assigned wants to preview and submit results
1. Tester who is assigned wants to browse forward or back in the plan or run tests non-sequentially.
1. Tester who is assigned wants to clear all fields on the current page.
1. Tester is assigned has submitted results for the currently displayed test but wants to discard the results and submit new results.
1. User who may or may not be assigned wants to report an issue with the test.
1. Tester who is assigned wants to flag that an issue is blocking execution of the test.
1. User who is not assigned wants to explore the runner without saving/submitting data.
1. Admin or tester wants to view associated blocking issue.
1. Admin or tester wants to ask system to remove block.

We do not need to create a dependency between reporting an issue and skipping. If the tester moves on to the next test without submitting results for the currently displayed test, that is fine. The test plan will simply remain incomplete.

I propose we can satisfy all the above uses cases with the following buttons:
1. "Review and Submit" button:
   * This button is hidden or disabled if the current user is not assigned to the test plan.
   * If the form is incomplete, it pops up a dialog titled "Error! Data Incomplete". It contains a list of incomplete fields and a button to continue, which places focus on the first incomplete field.
   * If the form is complete, it opens a dialog titled "Review Results". We need to discuss the design of this dialog for when there are conflicts. Otherwise, there are two buttons: "Save and Continue" and "Cancel". Pressing save and continue marks the test complete, saves the data, and displays the next test in the plan.
1. "Clear All" button: Opens alert dialog with two buttons: "Clear all data entered on this page" and "Cancel". This button is hidden or disabled if the current user is not assigned. It is disabled if the current user is assigned and all fields are in their initial state.
1. Next Test button: Available to all users all the time. If the current user is an assigned tester, this could save or cache data that has been entered. Alternatively, if saving or caching data that is not submitted doesn't fit in V1, then just prompt users asking if they want to discard data and continue or cancel.
1. Previous Test button: Available to all users all the time. If the current user is an assigned tester, this could save or cache data that has been entered. Alternatively, if saving or caching data that is not submitted doesn't fit in V1, then just prompt users asking if they want to discard data and continue or cancel.
1. "Back to Test Queue" button: Available to all users all the time. If the current user is an assigned tester, this could save or cache data that has been entered. Alternatively, if saving or caching data that is not submitted doesn't fit in V1, then just prompt users asking if they want to discard data and continue or cancel.
1. "Report Issue" button:
   * If the current user is assigned tester, opens a dialog asking "Does the issue you wish to report block recording accurate results for this test?". There are two buttons: "Yes, this is a blocking issue" and "No, testing can be completed without resolving this issue". After the question is answered, the GitHub new issue page is opened in a new tab.
   * If the user is not an assigned tester, the GitHub new issue page is opened in a new tab when the button is pressed.
   * When the new issue page is opened, the title and comment fields are prepopulated and appropriate labels are added. If the issue is a blocking issue, an appropriate label is added to the issue. We should work out the specifications for title, comment, and labels in a separate specification.
   * If the test is blocked, we may want to keep track of that status in the database and have a function for removing that status (stretch goal for later). For now, the fact that a test is blocked is something that testers and admin can communicate about outside the app.
   * Note that we are just pre-populating the GitHub issue so if the user changes their mind and wants to cancel, all the user need do is close the tab without submitting the issue.

I have a lot of different ideas about how we could handle the conflict display and resolution process for testers and admins. I think we need some design brain storming time for this. We should include engineering in the discussion so we can optimize for both experience and engineering simplicity.

Some possible options:
* Show differences in the review dialog
* Show differences in the runner after closing review
* Provide a function to copy differences to clipboard. This would be the same text format that would be used to populate an issue comment. But, the user could just paste into an editor and keep that open while reviewing what is recorded in the runner.

**Comment by zcorpan** [comment]
@mcking65 

> Based on last Wednesday's discussion about how to resolve conflicts, I think we need further discussion of process before designing raw test results and draft results pages. In fact, we may not need those pages at all. I am thinking we may want to build their functionality into the runner itself.

Adding `Agenda+` to discuss this today.

**Comment by mcking65** [comment]
I've incorporated my suggestions along with the changes with previous comment into the content from the wiki page. I'loosely marked the starting point of most changes with **MK**.

I didn't yet get to incorporating the changes to the review page I previously made with respect to calling out conflicts.

# Test Management
This page is accessible from the main menu. There are two sections in this page: 1) Initiate a Test Cycle and 2) Test Cycle Status

## 1. Initiate a Test Cycle 
Here, with the click of a button, a new Test Cycle initiation can be executed. This will take the user to another page where the Test Cycle can be configured. 

## 2. Test Cycle Status 
This is a table that displays the Test Cycles that have been initiated. For every Test Cycle there's a row in the table. This table has the following columns: 
   1. Test Cycle: It contains a link with the name given to the Test Cycle when it was configured. Clicking this will take the admin to the Test Cycle Summary Page.
   2. Date: The date in which the Test Cycle was initiated.
   3. **MK** Status: T Runs, N not started, P in-progress, R Raw Results, D Drafts Complete, F final Reports (Note: only show non-zero counts. p are runs where some raw results are pending, R are runs where all assigned testers have completed all tests in the run.)
   4. More Details: The cells of this table contain buttons/links to raw results, draft results and reviewed results.

**MK**: Per discussion, column 4 is optional in V1. People can get to the draft and final reports from the reports page and the raw results from the test queue page. In fact, we may never want to access raw results from here; that could be a bit complicated.

![Test Management](https://user-images.githubusercontent.com/1379244/77789855-11051980-7021-11ea-93e6-d3b85803b4e8.png)

# Initiating a Test Cycle
This page is generated from the Test Management Page. There are two main sections in this page: 1) Test Cycle Configuration, and 2) Test Plans Available

## 1. Test Cycle Configuration 
This is a table where the admin can choose a custom name for the cycle, initiate a cycle from a specific commit and choose specific Assistive Technology and Browser versions. This table has the following characteristics:

   1. Test Cycle name: This is a text field where a custom name can be added
   2. Specific commit: From a dropdown, the admin can choose a specific commit
   3. AT and Browser versions: This table has as many rows as AT and browser combinations are available. There's a row per AT and Browser combination. It has the following columns:
      1. **MK** Include: A checkbox (checked by default) that determines if the AT/Browser combo will be included in the cycle.
      1. Assistive Technology: Each cell has a dropdown that contains the available AT released versions such as: beta, stable, dev, etc
      1. Assistive Technology Version Number: This cell contains a text field where the admin can input a specific version number for the AT
      1. Browser: Each cell has a dropdown that contains the available browser stage released versions such as: beta, stable, etc
      1. Browser Version number: This cell contains a text field where the admin can input a specific version number for the Browser.

## 2. Test Plans Available
This section includes all available Test Plans in the system. The admin can decide to submit them all as part of a Test Cycle or choose a subset of these. Each Test Plan has its own table with all the available AT and Browser combinations available in the system as well as the testers assigned to execute certain Test plans with certain AT and Browser combinations. This table has the following characteristics:
 1. Test plan name: This is the title for the test plan. There's a checkbox next to it to select the test plan as part of the Test Cycle. All test plans are selected by default.
 2. The Test Plan table has the following columns:
    1. AT and Browser: This column lists the AT and Browser combinations available in the system. One combination is displayed by row. Each includes a checkbox to make a custom selection of AT/Browsers per Test Plan.
    2. Tester 1: This includes a dropdown where tester 1 is automatically selected by the system. The admin has the ability to override this selection choosing a different tester from the dropdown.
    3. Tester 2: Same characteristics of Tester 1.

**MK** Unless this is already built, we could  eliminate  tester assignment from this page and just have a list of all the test plans. It would be assumed that all plans would run with the set of browser/AT combinations checked off above.

At the bottom of this page, there's a UI component that contains a checkbox to check/uncheck all test plans and a button to initiate the Test Cycle.

![Initiating a Test Cycle](https://user-images.githubusercontent.com/1379244/77789911-2417e980-7021-11ea-8b12-855b14c7dcc8.png)

# Test Cycle Summary Page
This page shows a high level summary of the status of a Test Cycle and is accessible from the Test Management page by clicking the name of the Test Cycle from the Test Cycle Status table. There are two  sections in this page: 1) Summary of configurations, and 2) Test Plans included in the Test Cycle.

## 1. Summary of configurations
This section displays the summary of configurations chosen while initiating the Test Cycle: Test Cycle name, git commit, specific versions of AT and Browsers if any were selected.

**MK** Recommend that the cycle name and specific versions are edit fields and there is a save button that is enabled if any changes are made. If the user is not an admin, the edit fields are read-only.

## 2. Test Plans included in the Test Cycle
**MK** This section has a subsection for each test plan. It includes a heading with the test plan name and a table listing the runs for that plan.
1. The test plan heading includes the plan name and a status indicator:
   * 0% complete if not started
   * X% complete if testing in progress (# complete sets of raw results/total # required)
   * In Draft if all raw results are done and there is at least 1 draft
   * Final if all results are final
 1. The  table has the following columns:
   1. Select: checkbox for selecting each row for action. Checking the box in the column header checks or unchecks all rows.
   1. AT and Browser: This column lists the AT and Browser combinations included in the cycle.
   1. Progress by Tester: For each tester assigned, shows tester ID and Done/Total, e.g., Jon: 4 Done/16 Total
   1. Status: Not Started | Raw Results In-Progress | Raw Results Complete | Draft Report | Final Report

Below the table is an action button for adding testers to selected rows.

![Test Cycle Summary Page](https://user-images.githubusercontent.com/1379244/77789960-33973280-7021-11ea-8106-1ef114c513b0.png)

# **MK** Test Queue (formerly Executing Tests)

**MK** This page shows the Test Plans  defined by the admin when  a  Test Cycle was created. It includes a section for each active test cycle. Each cycle contains  a section for each AT/Browser combination included in the cycle. 
Each AT/Browser section includes a table listing the test plans, assigned testers with their progress, report status, and provides the ability to execute a test run and access to other relevant functions.

The default page content     depends on the role of the user, e.g., admin, tester, any other user. By default, testers will only see sections/plans they are qualified to run based on their profile.

At the top of the page,   there is a    view selector with options for:
1. Assigned to me (only available to testers)
1. Assigned + Up for Grabs ( Default for testers -- only available to testers)
1. All plans (Default for admins and any other user)
1. Plans with conflicting raw results
1. Plans blocked by issues (stretch goal for V1)
1. Plans with complete raw results
1. Plans with draft reports

The heading for each browser/AT section includes the AT name and version, Browser name and version, and a status indicator, e.g., "JAWS 2020 with Chrome 80 - 0% complete". The status is formulated in the same way it is formulated for the test cycle summary:
* 0% complete if not started
* X% complete if testing in progress (# complete sets of raw results/total # required)
* In Draft if all raw results are done and there is at least 1 draft
* Final if all results are final

The  table in each AT/Browser section has the following columns:
1. Test Plan: List available plans. The name is a link that opens the test run in the runner. This link has the following behaviors:
  * If the user is a tester and assigned to the plan run, and if the plan is still in raw state, checks if saved results from that tester exist, loads them if yes, opens the runner, displays first test with incomplete results. If all results are complete, and their no conflicts, shows first test. If all results are complete and there are conflicts, shows first test with conflicts.
  * If user is a qualified tester who is not assigned, and if the plan is still raw, opens a dialog to prompt if the user wishes to be assigned. If yes, opens the runner with save functions enabled. If no, opens with save functions disabled.
  * If user is admin or any other user, opens the runner with save functions disable.
1. Tester Progress: Shows whether testers are assigned and the progress of each tester:
  * If fewer than 2 testers are assigned, shows a bright "Tester Needed" indicator. If the user is a qualified tester, shows an ""Assign Me" button. If the user is an admin, shows an "Add Tester" button.
  * For each assigned tester, shows "TESTER_ID: n Done of T Total)
1. Report Status: Shows current report status:
  * Not Started: 0 testing is done.
  * In Progress: Some test data is saved, but there are required tests pending -- fewer than 2 testers assigned | 1 or more testers have completed less than 100% of tests.
  * N Conflicting Raw Results: In progress with at least 1 conflicting test result | All tests done and at least 1 conflicting test result
  * Raw Results: A link to the raw results rpoert
  * Draft Report: A link to the draft report
  * Final Report: A link to the final report.
1. Actions: Actions menu button that opens a menu with the following options:
  * Open as ... (has submenu): Visible only to admins, enables the admin to open the test run as one of the assigned testers who has saved data. There is an item  in the submenu for each assigned tester.
  * Unassign Me: shows only to assignee. Opens confirmation dialog. If data has been saved, warns that the user's data will be deleted for that run.)
  * Remove Assignee (has submenu): Shows only to admins. The submenu has an item for each assignee, e.g., "Remove Matt". Each item opens a confirmation; it warns if data will be deleted.
  * Delete My Results: Shows only to assigned tester. Opens a confirmation dialog that allows tester to delete all their results for the run)
  * Delete Results (has submenu): Shows only to admins. The submenu has an item for each assigned tester. Prompts for confirmation.
  * Add tester: shows only to admins. Opens a dialog that enables the admin to assign another tester. There is no limit to the number of testers that can be assigned.
  * Mark as draft: Shows only to admin when raw results are complete and there are no conflicts.
  * Mark as final: Shows only to admin when report is in draft state.
        
**MK** Once the tester has decided which test run to execute, the process looks pretty similar to what we currently have in the prototype with the following exceptions.

1. If the user is a tester, previously saved data is loaded when available.
1. A status appears at the top:
  * Complete: If  data is complete and  has been saved
  * Has Conflicts: If data is complete but is conflicting with results from other testers.
  * Blocked by Issue: If a blocking issue has been raised.
1. Users who are not testers can browse the test and raise issues. They can't save data.
1. The following action buttons are available:
  1. Previous Test
  1. Next test
  1. Raise issue
  1. Re-do (clears all fields after a prompt)
  1. Save and review (Opens review page)
  1. Save and close (returns to queue)
  
Note: the process for executing Tests is the same we have in the prototype.

![Executing Tests](https://user-images.githubusercontent.com/1379244/77789973-398d1380-7021-11ea-8823-a9079ffd7bf6.png)


# Raw Test Results

**MK** no longer needed.

# Draft Results

**MK** No longer needed

# Reports

This page provides all users to access to all draft and final report data from all test cycles. It enables users to:
1. See all reports  from a given cycle. Within that cycle:
  * See reports for a given AT
  * See  reports for a given browser
1. See all reports for a given AT, perhaps across cycles?
1. See all reports for a given browser, perhaps across cycles?
1. See all reports for a given plan
1. etc.

We need to discuss MVP.

**Comment by mcking65** [comment]
#### Format of test result diff

When results from 2 or more testers conflict, I've proposed that we enable users to view the details of those differences outside the system by providing a button to copy a formatted text string that describes the differences to the clipboard. Following is a format for that string. If it is copied to the clipboard, it could be plain text. If generated for GitHub, it could be markdown. We could also provide the ability to copy the markdown to the clipboard (I'd personally like that).

We should have a separate spec for everything that should go into a GitHub issue raised by the system. This format would work for the comment, but we need to specify the title, assignments, and labels as well.

------

Differences in results from tester1Id and tester2Id
When testing: {link to test case}
When using: {AT name/version} with {browser name/version}
For test: {test link}

Difference 1:
Testing command {command} for assertion {assertion}:
tester1Id: result
tester2Id: result

Difference 2:
Testing command {command} for assertion {assertion}:
tester1Id: result
tester2Id: result

**Comment by isaacdurazo** [comment]
Closing this discussion now. I will be creating new issues for new wireframe iterations
